Cannot reserve, resize nor emplace_back in std::vector<te::poly<Type>>

It's not possible to change internal or external size of std::vector that stores te::poly objects, nor emplace_back objects into it.

Example with reserving space in the vector:
```
#include "external/te/include/te.hpp"
#include <vector>

struct Drawable
{
    auto draw() -> REQUIRES(void, draw)
};

int main()
{
    std::vector<te::poly<Drawable>> drawables;
    drawables.reserve(1);
}
```
Its compilation fails with:
```
/usr/include/c++/7/type_traits:1162: error: invalid use of incomplete type ‘struct std::is_constructible<te::v1::poly<Drawable>, te::v1::poly<Drawable, te::v1::dynamic_storage, te::v1::static_vtable>&&>’
     struct __is_move_constructible_impl<_Tp, true>
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
When we change `reserve` to `resize`, we get the same error and additionally this one:
```
/usr/include/c++/7/bits/stl_construct.h:75: error: no matching function for call to ‘te::v1::poly<Drawable>::poly()’
     { ::new(static_cast<void*>(__p)) _T1(std::forward<_Args>(__args)...); }
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

When we change the code to test emplace_back with a potentially drawable object:
```
struct Circle
{
    void draw() {}
};

int main()
{
    std::vector<te::poly<Drawable>> drawables;
    drawables.emplace_back(Circle{});
}
```
we get the same error as in the `reserve` case.

**Environment**
OS: Ubuntu 17.10 x64
Compiler: GCC 7.2 x64

--- a/example/forward.cpp
+++ b/example/forward.cpp
@@ -25,7 +25,7 @@ struct Circle {
 // cpp
 void draw(te::poly<Drawable> const &drawable) { drawable.draw(std::cout); }
 
-void Drawable::draw(std::ostream& out) const {
+void Drawable::draw(std::ostream &out) const {
   te::call([](auto const &self, auto &out) { self.draw(out); }, *this, out);
 }
 

--- a/include/te.hpp
+++ b/include/te.hpp
@@ -196,11 +196,8 @@ class poly : detail::poly_base,
              public std::conditional_t<detail::is_complete<I>{}, I,
                                        detail::type_list<I> > {
  public:
-  template <class T,
-            class = std::enable_if_t<not std::is_convertible<T, poly>{} and
-                                     std::is_destructible<T>{} and
-                                     (std::is_copy_constructible<T>{} or
-                                      std::is_move_constructible<T>{})> >
+  template <class T, class = std::enable_if_t<
+                         not std::is_convertible<std::decay_t<T>, poly>{}> >
   constexpr poly(T &&t) noexcept
       : poly{std::forward<T>(t),
              detail::type_list<decltype(detail::requires__<I>(bool{}))>{}} {}
@@ -215,13 +212,16 @@ class poly : detail::poly_base,
       : poly{std::forward<T>(t),
              std::make_index_sequence<detail::mappings_size<I>()>{}} {}
 
-  template <class T, std::size_t... Ns>
+  template <class T, class T_ = std::decay_t<T>, std::size_t... Ns>
   constexpr poly(T &&t, std::index_sequence<Ns...>) noexcept
       : detail::poly_base{},
         vtable{std::forward<T>(t), vptr,
                std::integral_constant<std::size_t, sizeof...(Ns)>{}},
         storage{std::forward<T>(t), ptr} {
     static_assert(sizeof...(Ns) > 0);
+    static_assert(std::is_destructible<T_>{});
+    static_assert(std::is_copy_constructible<T>{} or
+                  std::is_move_constructible<T>{});
     (init<Ns + 1, std::decay_t<T> >(
          decltype(get(detail::mappings<I, Ns + 1>{})){}),
      ...);

--- a/test/te.cpp
+++ b/test/te.cpp
@@ -233,10 +233,11 @@ test should_support_movable_only_types_with_local_storage = [] {
 
 test should_support_containers = [] {
   std::vector<te::poly<Drawable>> drawables{};
+  drawables.reserve(42);
 
   drawables.push_back(Square{});
   drawables.push_back(Circle{});
-  drawables.push_back(Triangle{});
+  drawables.emplace_back(Triangle{});
 
   std::stringstream str{};
   for (const auto &drawable : drawables) {

